api_core/google/api_core/datetime_helpers.py

```diff
@@ -179,3 +179,109 @@ def to_rfc3339(value, ignore_zone=True):
         value = value.replace(tzinfo=None) - value.utcoffset()
 
     return value.strftime(_RFC3339_MICROS)
+
+
+class TimestampWithNanoseconds(datetime.datetime):
+    """Track nanosecond in addition to normal datetime attrs.
+
+    nanosecond can be passed only as a keyword argument.
+    """
+    __slots__ = ('_nanosecond',)
+
+    # pylint: disable=arguments-differ
+    def __new__(cls, *args, **kw):
```

Comment:
Why are we overriding `__new__` instead of `__init__`? @tseaver?

Reply:
@jonparrott Try deriving from `datetime.datetime` and using `__init__`:  the majyk happens inside `__new__`, so we have to override it:

```python
# override_datetime___init__.py
import datetime

class MyDateTime(datetime.datetime):

    __slots__ = ('foo',)

    def __init__(self, *args, **kw):
        foo = kw.pop('foo', None)
        super(MyDateTime, self).__init__(*args, **kw)
        self.foo = foo

mdt = MyDateTime(2018, 2, 28, 23, 11, 57, foo='bar')
```

```bash
$ python override_datetime___init__.py
Traceback (most recent call last):
  File "override_datetime___init__.py", line 13, in <module>
    mdt = MyDateTime(2018, 2, 28, 23, 11, 57, foo='bar')
TypeError: 'foo' is an invalid keyword argument for this function
```